web/src/helpers/api.js

```diff
@@ -0,0 +1,86 @@
+import * as request from 'd3-request';
+import { sha256 } from 'js-sha256';
+import Cookies from 'js-cookie';
+
+import { isLocalhost, isProduction } from './environment';
+import thirdPartyServices from '../services/thirdparty';
+
+function isRemoteParam() {
+  return (new URLSearchParams(window.location.search)).get('remote') === 'true';
+}
+
+// Use local endpoint only if ALL of the following conditions are true:
+// 1. The app is running on localhost
+// 2. The `remote` search param hasn't been explicitly set to true
+// 3. Document domain has a non-empty value
+function isUsingLocalEndpoint() {
+  return isLocalhost() && !isRemoteParam() && document.domain !== '';
+}
+
+export function getEndpoint() {
+  return isUsingLocalEndpoint() ? 'http://localhost:9000' : 'https://api.electricitymap.org';
+}
+
+export function protectedJsonRequest(path) {
+  const url = getEndpoint() + path;
+  const token = isUsingLocalEndpoint() ? 'development' : ELECTRICITYMAP_PUBLIC_TOKEN;
+  const timestamp = new Date().getTime();
+
+  return new Promise((resolve, reject) => {
+    request.json(url)
+      .header('electricitymap-token', Cookies.get('electricitymap-token'))
+      .header('x-request-timestamp', timestamp)
+      .header('x-signature', sha256(token + path + timestamp))
+      .get(null, (err, res) => {
+        if (err) {
+          reject(err);
+        } else if (!res || !res.data) {
+          reject(new Error(`Empty response received for ${url}`));
+        } else {
+          resolve(res.data);
+        }
+      });
+  });
+}
+
+export function textRequest(path) {
+  const url = getEndpoint() + path;
+
+  return new Promise((resolve, reject) => {
+    request.text(url).get(null, (err, res) => {
+      if (err) {
+        reject(err);
+      } else if (!res) {
+        reject(new Error(`Empty response received for ${url}`));
+      } else {
+        resolve(res);
+      }
+    });
+  });
+}
+
+function trackError(err, appState) {
+  console.error(`Error Caught! ${err}`);
+  thirdPartyServices.reportError(err);
+  thirdPartyServices.ga('event', 'exception', { description: err, fatal: false });
+  thirdPartyServices.track('error', Object.assign({}, appState, { name: err.name, stack: err.stack }));
+}
+
+export function handleConnectionReturnCode(err, appState = {}) {
+  if (err) {
+    if (err.target) {
+      // Avoid catching HTTPError 0
+      // The error will be empty, and we can't catch any more info
+      // for security purposes
+      // See http://stackoverflow.com/questions/4844643/is-it-possible-to-trap-cors-errors
+      if (err.target.status) {
```

Comment:
can we also make sure not to catch errors 5xx? They appear a lot on Sentry as soon as the server experiences some downtime.

Reply:
Yes, it makes sense not to track them - I made the change in https://github.com/tmrowco/electricitymap-contrib/pull/2388/commits/507190a5fb9d7d1c228d91e96daaa855a01dee0b not to track them anymore and I also added tracking for zone history fetches in that commit which seems to have been missing from the code so far :thinking: